src/bridge/events.rs

```diff
@@ -159,494 +160,455 @@ fn unpack_color(packed_color: u64) -> Color4f {
     }
 }
 
-fn parse_array(array_value: &Value) -> Result<&[Value]> {
-    if let Value::Array(content) = array_value {
-        Ok(&content[..])
-    } else {
-        Err(EventParseError::InvalidArray(array_value.clone()))
-    }
+fn pop_value(values: &mut Vec<Value>) -> Result<Value> {
+    values.pop().ok_or(EventParseError::InvalidEventFormat)
 }
 
-fn parse_map(map_value: &Value) -> Result<&[(Value, Value)]> {
-    if let Value::Map(content) = map_value {
-        Ok(&content[..])
-    } else {
-        Err(EventParseError::InvalidMap(map_value.clone()))
+macro_rules! extract_values {
+    ([$($i: ident),+], $values: expr $(,)?) => {
+        let mut __values = $values;
+        if __values.len() != {
+            #[allow(dead_code, non_camel_case_types)]
+            enum Idents { $($i,)+ __CountIdentLast }
+
+            Idents::__CountIdentLast as usize
+        } {
+            return Err(EventParseError::InvalidEventFormat);
+        }
+
+        let __values_ref = &mut __values;
+        $(
+            let $i = pop_value(__values_ref)?;
+        )+
     }
 }
 
-fn parse_string(string_value: &Value) -> Result<&str> {
-    if let Value::String(content) = string_value {
-        content.as_str().ok_or_else(|| EventParseError::InvalidString(string_value.clone()))
-    } else {
-        Err(EventParseError::InvalidString(string_value.clone()))
-    }
+fn parse_array(array_value: Value) -> Result<Vec<Value>> {
+    array_value.try_into().map_err(EventParseError::InvalidArray)
 }
 
-fn parse_u64(u64_value: &Value) -> Result<u64> {
-    if let Value::Integer(content) = u64_value {
-        Ok(content.as_u64().ok_or_else(|| EventParseError::InvalidU64(u64_value.clone()))?)
-    } else {
-        Err(EventParseError::InvalidU64(u64_value.clone()))
-    }
+fn parse_map(map_value: Value) -> Result<Vec<(Value, Value)>> {
+    map_value.try_into().map_err(EventParseError::InvalidMap)
 }
 
-fn parse_i64(i64_value: &Value) -> Result<i64> {
-    if let Value::Integer(content) = i64_value {
-        Ok(content.as_i64().ok_or_else(|| EventParseError::InvalidI64(i64_value.clone()))?)
-    } else {
-        Err(EventParseError::InvalidI64(i64_value.clone()))
+fn parse_string(string_value: Value) -> Result<String> {
```

Comment:
This could use https://docs.rs/rmpv/0.4.4/rmpv/enum.Value.html#impl-TryFrom%3CValue%3E